automated/linux/lkp/lkp.sh

```diff
@@ -0,0 +1,347 @@
+			-S | --skip-test-install)
+				#${var,,} make variable to lowercase; require:bash >=4.0
+				if [ "${2,,}" = 'true' ] || [ "${2,,}" = 'false' ] ; then
+					SKIP_TEST_INSTALL="${2,,}"
+					shift; shift;
+				else
+					error_msg "'--skip-test-install' value must be [true|false]"
+					exit 1
+				fi
+				;;
+			-t | --test)
+				read -r -a TEST_QUEUE <<< "$2"
+				shift; shift;
+				;;
+
+			-T | --atomic-job)
+				read -r -a ATOMIC_JOBS <<< "$2"
+				shift; shift;
+				;;
+
+			-V | --prog-version)
+				if [[ "${2}" != '' ]]; then
+					TEST_PROG_VERSION="${2}"
+				fi
+				shift; shift;
+				;;
+
+			-h | --help)
+				usage
+				exit 1
+				;;
+			*)
+				error_msg "'${param}' is not a defined option"
+				exit 1;
+				;;
+			esac
+	done
+}
+
+#--- Install lkp from git repository unless otherwise specified
+# - INPUT: 	<None>
+# - SET:	LKP_PATH
+# - MODIFY: 	PWD
+install()
+{
+	if [[ "${SKIP_INSTALL}" = "false" ]]; then
+		info_msg "LKP installation ..."
+		#avoid lkp installation to prompt user creation
+		useradd lkp -m -p lkp || [[ "$?" == "9" ]]
+		[[ -n "${TEST_GIT_URL}" ]] && pkgs+=" git"
+		pkgs+=" make"
+		pkgs+=" build-essential" #avoid aarch64 incorrect lkp installation
+		install_deps "${pkgs}"
+		get_test_program "${TEST_GIT_URL}" "${TEST_DIR}" "${TEST_PROG_VERSION}" "${TEST_PROGRAM}"
+
+		cd "${TEST_DIR}" || exit 1
+
+		eval make install "${QUIET}"
+		eval lkp install "${QUIET}"
+	else
+		info_msg "LKP installation skipped"
+	fi
+
+	LKP_PATH=$(readlink -f "$(which lkp)" 2>/dev/null | sed 's:bin.*::')
+}
+
+#--- Install single atomic-jobs dependencies unless otherwise specified
+# - INPUT: 	[ext] ATOMIC_JOBS[@]
+# - SET:	<None>
+# - MODIFY: 	<None>
+install_atomic_jobs_dependencies()
+{
+	if [[ "${SKIP_TEST_INSTALL}" = "false" ]]; then
+		info_msg "--Installing atomic jobs..."
+		for job in "${ATOMIC_JOBS[@]}"; do
+			echo "lkp install ${job} ${QUIET}"
+			lkp install "${job}" "${QUIET}"
+		done
+	else
+		info_msg "--Skip atomic jobs dependencies installation"
+	fi
+}
+#--- Create atomic-jobs from test with $ lkp split-job
+# - INPUT: 	[ext] test
+# - SET: 	ATOMIC_JOBS
+# - MODIFY	<None>
+create_atomic_jobs()
+{
+	info_msg "--Creating atomic-jobs for ${test}..."
+	readarray -t ATOMIC_FILES <<< "$(lkp split-job "${LKP_PATH}"/jobs/"${test}".yaml | cut -f 3 -d ' ')"
+
+}
+
+#--- Get atomic job exit code from result files
+# - INPUT: 	$1 =  Name of .yaml file run by run_atomic_job()
+# - SET:	<None>
+# - MODIFY: 	<None>
+report_atomic_job()
+{
+	info_msg "++Getting result of ${job}..."
+
+	local REPORT_CODE=''
+
+	if [[ ! -e "${TEST_DIR}/working/result" ]] ; then
+		info_msg "${1} didn't produce any results."
+		REPORT_CODE="1"
+	else	# gather atomic_job exit status accordingly test-name
+		case "${test}" in
+
+		sysbench-cpu)
+			# 'grep' return 1 if !found or 2 with error (file does not exists)
+			# 'exits_fail' appears in the output only if sysbench-cpu test failed
+			grep -r "exit_fail" "${TEST_DIR}/working/result" >/dev/null
```

Comment:
How is the `result` file generated? Can we rely on the exit code from `lkp run test`?  Maintaining test specific result parsing here sounds too much. 

IMHO, lkp.sh should be just a wrapper of a set of lkp commands. The current patch looks good for this part. For sending lkp results to lava, I think it will be ideal if we can mange to do the below two steps.

1) generate overall result for a test run by checking `lkp run test` exist code
2) convert lkp result(e.g. /lkp/result/sysbench-cpu/20000-100%-300s/82cd3896aa8e/debian/defconfig/gcc-8/5.10.0-1029-oem/0/sysbench-cpu.json) to `result.txt` for sending to lava.

Reply:
> How is the `result` file generated? 

`result` is a simlink generated by lkp at the end of the run and links to the last result-folder.
Alternatively there is a lkp built-in command: `lkp result <test>`

> Can we rely on the exit code from `lkp run test`? Maintaining test specific result parsing here sounds too much.

> 1. generate overall result for a test run by checking `lkp run test` exist code

`lkp run <test>` exit code is not useful, because it refers to lkp suite command and not about `<test>`.
You can test it simply running this:
```
git clone https://github.com/intel/lkp-tests.git
cd lkp-tests
sudo make install
sudo lkp install
sudo lkp split-job jobs/sysbench-cpu.yaml
sudo lkp run ./sysbench-cpu-20000-100%-300s.yaml
echo $?
```
Since `lkp install ./sysbench-cpu-20000-100%-300s.yaml` wasn't run, the test should be fail because of incomplete dependencies 

> IMHO, lkp.sh should be just a wrapper of a set of lkp commands. The current patch looks good for this part. For sending lkp results to lava, I think it will be ideal if we can mange to do the below two steps.

> 2. convert lkp result(e.g. /lkp/result/sysbench-cpu/20000-100%-300s/82cd3896aa8e/debian/defconfig/gcc-8/5.10.0-1029-oem/0/sysbench-cpu.json) to `result.txt` for sending to lava.

As I've understood, the actual `<test>` (named `atomic job` by lkp-docs) exit status need to be retrieved from output results, which are different from test to test. I tried a quick "fail-run" with ebizzy and hackbench tests, but each of them need a individual and specific parsing of the output folder to find out exit status. 
Actually the output is parsed only to retrieve exit_status, so no metrics is searched for nor is sended to Lava.
But surely this function can be moved to another file/script to make everything more modular if you think so.

Reply:
I just tested. You are right about `lkp run test` exist code. The first point here is a generic way should be used for getting exist code. If we are grepping `exit_fail` in lkp sub-test result folder, why this is different by test/sub-test? I understand the path to sub-test result folder can be retrieved dynamically using `lkp result test`, right? Maybe, the problem here is that it may return more than one paths, but we should be able work around this by removing `/lkp/result/test' folder before each sub-test run as results should be saved in lava already.

The second point here is that test metrics should be saved in lava if it is generated. Looks like it is possible to construct path to metric file as well. Something like `$(lkp result test-case-name)/test-case-name.json`

Reply:
test test test

Reply:
> I just tested. You are right about lkp run test exist code. The first point here is a generic way should be used for getting exist code. If we are grepping exit_fail in lkp sub-test result folder, why this is different by test/sub-test? 

A result path example is: `/lkp/result/sysbench-cpu/20000-100%-300s/debian/debian/defconfig/gcc-8/4.19.0-16-arm64/`
with `sysbench_cpu` as test, `20000-100%-300s` is <atomic-job> (test configuration).
Inside that folder there are standard lkp ouput file: `avg.csv, avg.json, desc.yaml, matrix.csv, matrix.json, stddev.csv, stddev.json` which should be comparable output with different tests run.
There is also at least one folder named with an increasing progressive integer (eg:  `0`, `1`, ...) which contains the specific test output.
From what I've saw from running `hackbench, ebizzy` and `sysbench-cpu` (successful and failed run) and comparing them, there is no common file or line to search for to understand exit status. Or at least, I've not understood there is one.

>  I understand the path to sub-test result folder can be retrieved dynamically using `lkp result test`, right? Maybe, the problem here is that it may return more than one paths, but we should be able work around this by removing `/lkp/result/test' folder before each sub-test run as results should be saved in lava already.

You are right with `lkp result test` which could return more than one path, but I never use it:
[here](https://github.com/Linaro/test-definitions/blob/2e6ed940be8708a6d33fd0ea8cb4401894fd587b/automated/linux/lkp/lkp.sh#L244) you can see that I used `result` inside `working directory`: this is a link created by lkp in the current directory;
[after](https://github.com/Linaro/test-definitions/blob/2e6ed940be8708a6d33fd0ea8cb4401894fd587b/automated/linux/lkp/lkp.sh#L277) I delete the link to avoid a successive failed run can access previous results. A successful run would overwrite the link and any problem shouldn't occur, but an incomplete run could not overwrite and trying to get old results.

> The second point here is that test metrics should be saved in lava if it is generated. Looks like it is possible to construct path to metric file as well. Something like $(lkp result test-case-name)/test-case-name.json

My job is related to an University exam and my goal is to write this wrapper, so I'm quite agnostic about test results and metrics; to retrieve the latter, some more time is required because I need to study 'sysbench-cpu' more. It could be useful if @roxell or @nareshkamboju tell me about what they want to know from the test.

Reply:
> > The second point here is that test metrics should be saved in lava if it is generated. Looks like it is possible to construct path to metric file as well. Something like $(lkp result test-case-name)/test-case-name.json
> 
> My job is related to an University exam and my goal is to write this wrapper, so I'm quite agnostic about test results and metrics; to retrieve the latter, some more time is required because I need to study 'sysbench-cpu' more. It could be useful if @roxell or @nareshkamboju tell me about what they want to know from the test.

We shouldn't need to study the 'sysbench-cpu' more, the outputs should be generated in a json file. In the directory that @chase-qi shows above you can dig out the results via 'lkp result test-case-name'.
So we should just need to parse that file so LAVA understands it.

Reply:
Oh yes, sorry I misunderstood output. A couple of day and I'll deliver the new patch. 

Thank you for your patience.

Reply:
Hi @roxell and @chase-qi,
I rewrite "report_atomic_job" function and now it retrieves and parse main <test>.json correctly.
Now everything is also reported to result.txt file so that Lava should use test data correctly.
I hope that now the patch is fine.

Reply:
Thanks @Magnani95. It looks good to me now.